libraries/rush-lib/src/logic/operations/OperationStateFile.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import { FileSystem, InternalError, JsonFile } from '@rushstack/node-core-library';
+import { RushConstants } from '../RushConstants';
+
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+
+export interface IOperationStateFileOptions {
+  rushProject: RushConfigurationProject;
+  phase: IPhase;
+}
+
+export interface IOperationStateJson {
+  durationInSecondsWithoutCache: number;
+}
+
+export class OperationStateFile {
+  private readonly _rushProject: RushConfigurationProject;
+  private _filename: string;
+
+  public constructor(options: IOperationStateFileOptions) {
+    const { rushProject, phase } = options;
+    this._rushProject = rushProject;
+    this._filename = OperationStateFile.getFilename(phase, rushProject);
+  }
+
+  public static getFilename(phase: IPhase, project: RushConfigurationProject): string {
+    const relativeFilename: string = OperationStateFile.getFilenameRelativeToProjectRoot(phase);
+    return path.join(project.projectFolder, relativeFilename);
+  }
+
+  public static getFilenameRelativeToProjectRoot(phase: IPhase): string {
```

Comment:
Is it useful to have this as a separate function?

Reply:
The reason is `ProjectBuildCache` expects relative paths. For example, `.rush/temp/operation/build/state.json` is given. The <package_name>-<phase_name>-<identifier>.log would be

```
======= BEGIN PROCESS INPUT ======
.rush/temp/operation/_phase_build/state.json  <-- relative
dist/index.js
dist/index.d.ts
```

If an absolute path is given, it works. but the log would be

```
======= BEGIN PROCESS INPUT ======
/Users/<user_name>/<repo_root>/<project_folder>/.rush/temp/operation/_phase_build/state.json  <-- absolute
dist/index.js
dist/index.d.ts
```

No big deal tho